prover/protocol/compiler/plonkinwizard/compile.go

```diff
@@ -0,0 +1,155 @@
+	checkPublicInputs(comp, ctx)
+
+	comp.RegisterProverAction(round, &circAssignment{context: ctx})
+}
+
+// checkPublicInputs adds the constraints ensuring that the public inputs are
+// consistent with the one of the PlonkCtx.
+func checkPublicInputs(comp *wizard.CompiledIOP, ctx *context) {
+	comp.InsertGlobal(
+		ctx.Q.GetRound(),
+		ifaces.QueryIDf("%v_PUBLIC_INPUTS", ctx.Q.ID),
+		sym.Sub(ctx.Q.Data, ctx.PlonkCtx.ConcatenatedTinyPIs(ctx.Q.Data.Size())),
+	)
+}
+
+// checkActivators adds the constraints checking the activators are well-set w.r.t
+// to the circuit mask column. See [compilationCtx.Columns.Activators].
+func checkActivators(comp *wizard.CompiledIOP, ctx *context) {
+
+	var (
+		openings   = make([]query.LocalOpening, ctx.Q.GetMaxNbCircuitInstances())
+		mask       = ctx.Q.Selector
+		offset     = utils.NextPowerOfTwo(ctx.Q.GetNbPublicInputs())
+		activators = ctx.PlonkCtx.Columns.Activators
+		round      = activators[0].Round()
+	)
+
+	for i := range openings {
+		openings[i] = comp.InsertLocalOpening(
+			round,
+			ifaces.QueryIDf("%v_ACTIVATOR_LOCAL_OP_%v", ctx.Q.ID, i),
+			column.Shift(mask, i*offset),
+		)
+	}
+
+	comp.RegisterProverAction(ctx.Q.GetRound(), &assignSelOpening{context: ctx})
+	comp.RegisterVerifierAction(ctx.Q.GetRound(), &checkActivatorAndMask{context: ctx})
+	ctx.SelOpenings = openings
+}
+
+// checkSelectorAndData adds the constraints ensuring that the selector and the data
+// column are well-set w.r.t. to the circuit mask column.
+func checkSelectorAndData(comp *wizard.CompiledIOP, ctx *context) {
+
+	var (
+		round       = ctx.Q.GetRound()
+		nbPub       = ctx.Q.GetNbPublicInputs()
+		nbPubPadded = utils.NextPowerOfTwo(nbPub)
+		fullSize    = ctx.Q.Data.Size()
+	)
+
+	ctx.CircuitMask = ctx.Q.CircuitMask
+
+	// As the circuit mask is an optional parameter we have to account
+	// for the case where it is not provided by the user.
+	if ctx.CircuitMask == nil {
+
+		circMaskVal := make([]field.Element, ctx.Q.Data.Size())
+
+		for i := 0; i < fullSize; i += nbPubPadded {
+			for k := 0; k < nbPub; k++ {
+				circMaskVal[i+k] = field.One()
+			}
+		}
+
+		ctx.CircuitMask = comp.InsertPrecomputed(
+			ifaces.ColIDf("%v_CIRCMASK", ctx.Q.ID),
+			smartvectors.NewRegular(circMaskVal),
+		)
+	}
+
+	commonconstraints.MustBeActivationColumns(comp, ctx.Q.Selector)
+
+	// This query ensures that sel[i] - sel[i+1] == 1 => mask[i] - mask[i+1] == -1
+	// Note, that since sel is constrained to be an activation column the difference
+	// is already constrained to never be "-1", this allows simplifying the constraint
+	// as follows. The constraint only works if the number of public inputs is not
+	// exactly a power of two but in this case, the implemented approach does not
+	// work.
+	comp.InsertGlobal(
+		round,
+		ifaces.QueryIDf("%v_MASK_DEC_WHEN_SEL_DEC", ctx.Q.ID),
+		sym.Mul(
+			sym.Sub(ctx.Q.Selector, column.Shift(ctx.Q.Selector, 1)),
+			sym.Sub(ctx.CircuitMask, column.Shift(ctx.CircuitMask, 1), -1),
```

Comment:
For me it seems the constant should be `1` here? Otherwise, this would only hold if we have 0->1 transition?

Reply:
That's actually intended, 

0 -> 1, indicates the end of a mask period (e.g. a new plonk instance)
And I want the selector to cancel exactly just before the next instance. It's a bit over-constrained, I don't think it would be a big issue if we allowed sel: 1 -> 0 when mask: 0 -> * but this way there is only one valid assignment for sel which gives me more confidence.